library/waku_thread/inter_thread_communication/waku_thread_request.nim

```diff
@@ -28,22 +29,79 @@ type RequestType* {.pure.} = enum
 type InterThreadRequest* = object
   reqType: RequestType
   reqContent: pointer
+  callback: WakuCallBack
+  userData: pointer
+  doneSignal: ThreadSignalPtr
 
 proc createShared*(
-    T: type InterThreadRequest, reqType: RequestType, reqContent: pointer
+    T: type InterThreadRequest,
+    reqType: RequestType,
+    reqContent: pointer,
+    callback: WakuCallBack,
+    userData: pointer,
 ): ptr type T =
   var ret = createShared(T)
   ret[].reqType = reqType
   ret[].reqContent = reqContent
+  ret[].callback = callback
+  ret[].userData = userData
+  ret[].doneSignal = ThreadSignalPtr.new().tryGet()
   return ret
 
-proc process*(
-    T: type InterThreadRequest, request: ptr InterThreadRequest, waku: ptr Waku
-): Future[Result[string, string]] {.async.} =
-  ## Processes the request and deallocates its memory
+proc handleRes[T: string | void](
+    res: Result[T, string], request: ptr InterThreadRequest
+) =
+  ## Handles the Result responses, which can either be Result[string, string] or
+  ## Result[void, string].
+
   defer:
-    deallocShared(request)
+    let fireSyncRes = request[].doneSignal.fireSync()
+    if fireSyncRes.isErr():
+      let msg = "libwaku error: " & $fireSyncRes.error
+      request[].callback(
+        RET_ERR, unsafeAddr msg[0], cast[csize_t](len(msg)), request[].userData
+      )
+      return
+    if fireSyncRes.get() == false:
+      let msg = "libwaku error: couldn't fireSync in time"
+      request[].callback(
+        RET_ERR, unsafeAddr msg[0], cast[csize_t](len(msg)), request[].userData
+      )
+      return
+
+  if res.isErr():
+    foreignThreadGc:
+      let msg = "libwaku error: " & $res.error
+      request[].callback(
+        RET_ERR, unsafeAddr msg[0], cast[csize_t](len(msg)), request[].userData
+      )
+    return
+
+  foreignThreadGc:
+    var msg: cstring = ""
+    when T is string:
+      msg = res.get().cstring()
+    request[].callback(
+      RET_OK, unsafeAddr msg[0], cast[csize_t](len(msg)), request[].userData
+    )
+  return
 
+proc deallocOnDone*(T: type InterThreadRequest, request: ptr InterThreadRequest) =
```

Comment:
I wonder if we could make this private and invoke it within a `defer` block inside the `handleRes` proc.
```suggestion
proc deallocOnDone(T: type InterThreadRequest, request: ptr InterThreadRequest) =
```

Reply:
Hm not sure how at the moment. I'm using this deallocOnDone() function as a way to get the functions in libwaku.nim to 'wait' until there's a result and dispose of the waku request afterwards :thinking: